### PR TITLE
Rack attack

### PR DESCRIPTION
**Pull request by nbassagoda**
#### :pushpin: Notes:

- By default, rack-attack won't perform any blocking or throttling, until you specifically tell it what to protect against by configuring some rules. Therefore I added an example config that will prevent most of bad requests and can then be modified to serve their specific requirements.

---

#### :heavy_check_mark:Tasks:

- Add gem: Rack Attack

- Add example config for this to work out of the box

---

@loopstudio/ruby-devs

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -70,3 +70,6 @@ group :test do
 end
 
 gem 'tzinfo-data', platforms: %i[mingw mswin x64_mingw jruby]
+
+# Rack middleware for blocking & throttling abusive requests
+gem 'rack-attack'
```
Comment (mconiglio) [inline Gemfile:75]: this should be under the `# WebServer` section of the gemfile

Comment (nbassagoda) [reply Gemfile]: Done 🙌

File: config/initializers/rack_attack.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  # passwords to see if any credentials match.
+  #
+  # Another common method of attack is to use a swarm of computers with
+  # different IPs to try brute-forcing a password for a specific account.
+
+  # Throttle POST requests to /login by IP address
+  #
+  # Key: "rack::attack:#{Time.now.to_i/:period}:logins/ip:#{req.ip}"
+  throttle('logins/ip', limit: 5, period: 20.seconds) do |req|
+    if req.path == '/api/v1/sessions' && req.post?
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:44]: did you test if this works properly? I think this should be `/api/v1/users/sign_in`

File: config/initializers/rack_attack.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  # Throttle POST requests to /login by email param
+  #
+  # Key: "rack::attack:#{Time.now.to_i/:period}:logins/email:#{normalized_email}"
+  #
+  # Note: This creates a problem where a malicious user could intentionally
+  # throttle logins for another user and force their login requests to be
+  # denied, but that's not very common and shouldn't happen to you. (Knock
+  # on wood!)
+  throttle("logins/email", limit: 5, period: 20.seconds) do |req|
+    if req.path == '/api/v1/sessions' && req.post?
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:58]: same

File: Gemfile [hunk 4]
```diff
@@ -9,6 +9,9 @@ gem 'puma', '~> 5.0'
 gem 'rack', '~> 2.2.3'
 gem 'rack-cors', '~> 1.1.1'
 
+# Rack middleware for blocking & throttling abusive requests
+gem 'rack-attack'
+
```
Comment (elestu) [inline Gemfile:14]:
```suggestion
gem 'rack-attack'
```

File: config/initializers/rack_attack.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      # Normalize the email, using the same logic as your authentication process, to
+      # protect against rate limit bypasses. Return the normalized email if present, nil otherwise.
+      req.params['email'].to_s.downcase.gsub(/\s+/, "").presence
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:61]: is this how devise does it?

File: config/initializers/rack_attack.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  # Throttle POST requests to /login by email param
+  #
+  # Key: "rack::attack:#{Time.now.to_i/:period}:logins/email:#{normalized_email}"
+  #
+  # Note: This creates a problem where a malicious user could intentionally
+  # throttle logins for another user and force their login requests to be
+  # denied, but that's not very common and shouldn't happen to you. (Knock
+  # on wood!)
+  throttle("logins/email", limit: 5, period: 20.seconds) do |req|
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:57]: single quotes

**Comment by mconiglio** [comment]
what's the error body when rack-attack makes the requests fail? does it respect the errors format we are using?

File: config/initializers/rack_attack.rb (hunk 5 above)
Comment (nbassagoda) [reply config/initializers/rack_attack.rb]: @mconiglio I believe device does not implement any normalization logic by default and we haven't configured any either, that's why I simply left this simple check that was the one that was by default in the example

**Comment by nbassagoda** [comment]
> what's the response body when rack-attack makes the requests fail? does it respect the errors format we are using?

By default, it returns an HTTP 429 for throttled responses with no body I believe. The response can be customized though

File: config/initializers/rack_attack.rb (hunk 5 above)
Comment (mconiglio) [reply config/initializers/rack_attack.rb]: I think devise does normalize the emails by default, at least with a minimum logic like downcasing as mails are never stored with upper case characters. What I'm not sure is if it does something else, do you mind checking so we have the exact same logic matching here?

**Comment by mconiglio** [comment]
> > what's the response body when rack-attack makes the requests fail? does it respect the errors format we are using?
> 
> By default, it returns an HTTP 429 for throttled responses with no body I believe. The response can be customized though

could you customize the response body to match our error responses?

File: config/initializers/rack_attack.rb (hunk 5 above)
Comment (mconiglio) [reply config/initializers/rack_attack.rb]: it seems that under the hood it's essentially doing `email.downcase.strip` [here](https://github.com/heartcombo/devise/blob/eed641d2bea11839ab13e943660da41cad14314d/lib/devise/models/authenticatable.rb#L211) and [here](https://github.com/heartcombo/devise/blob/eed641d2bea11839ab13e943660da41cad14314d/lib/devise/models/authenticatable.rb#L215)

**Comment by nbassagoda** [comment]
> > > what's the response body when rack-attack makes the requests fail? does it respect the errors format we are using?
> > 
> > 
> > By default, it returns an HTTP 429 for throttled responses with no body I believe. The response can be customized though
> 
> could you customize the response body to match our error responses?

I just pushed this and also added some tests

File: config/initializers/rack_attack.rb (hunk 5 above)
Comment (nbassagoda) [reply config/initializers/rack_attack.rb]: Just changed it in order to match this. Thanks!

File: config/initializers/rack_attack.rb [hunk 7]
```diff
@@ -70,9 +70,11 @@ class Rack::Attack
...
-  # self.throttled_response = lambda do |env|
-  #  [ 503,  # status
-  #    {},   # headers
-  #    ['']] # body
-  # end
+  self.throttled_response = lambda do |env|
+    [
+      429,
+      {'Content-Type' => 'application/json'},
+      [{errors: ["Throttle limit reached"]}.to_json]
```
Comment (lucaleivaloop) [inline config/initializers/rack_attack.rb:77]: This should be i18n.

File: spec/config/initializers/rack_attack_spec.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      let(:params) do
+        {
+          user: {
+            email: "wrong+#{SecureRandom.hex(5)}@email.com",
+            password: 'wrong_password'
+          }
+        }
+      end
+
+      let(:headers) { { REMOTE_ADDR: "1.2.3.4" } }
```
Comment (nbassagoda) [inline spec/config/initializers/rack_attack_spec.rb:55]:
```suggestion
      let(:headers) { {REMOTE_ADDR: "1.2.3.4"} }
```

File: Gemfile [hunk 9]
```diff
@@ -8,7 +8,7 @@ gem 'rails', '~> 6.0.3'
 gem 'puma', '~> 5.0'
 gem 'rack', '~> 2.2.3'
 gem 'rack-cors', '~> 1.1.1'
```
Comment (mwolman) [inline Gemfile:10]:
```suggestion
gem 'rack-attack', '~> 6.2.2'
```

File: Gemfile [hunk 10]
```diff
@@ -8,7 +8,7 @@ gem 'rails', '~> 6.0.3'
...
-
+gem 'rack-attack'
```
Comment (mwolman) [inline Gemfile:11]:
```suggestion
gem 'rack-cors', '~> 1.1.1'

```

File: config/locales/en.yml [hunk 11]
```diff
@@ -17,6 +17,7 @@ en:
         other: Users
   errors:
     server: 'An error ocurred'
+    network_throttling: 'Throttle limit reached'
```
Comment (mwolman) [inline config/locales/en.yml:20]: Missing es translation

File: spec/config/initializers/rack_attack_spec.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        user: {
+          email: 'wrong@email.com',
+          password: 'wrong_password'
+        }
+      }
+    end
+
+    let(:headers) { {} }
+
+    context 'number of requests is lower than the limit' do
```
Comment (mwolman) [inline spec/config/initializers/rack_attack_spec.rb:22]:
```suggestion
    context 'when the number of requests is lower than the limit' do
```

Comment (mwolman) [reply spec/config/initializers/rack_attack_spec.rb:22]: Same comment for the other `context` blocks
https://rspec.rubystyle.guide/#context-descriptions

File: spec/config/initializers/rack_attack_spec.rb [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+      it do
```
Comment (mwolman) [inline spec/config/initializers/rack_attack_spec.rb:23]:
```suggestion
      specify do
```

Comment (mwolman) [reply spec/config/initializers/rack_attack_spec.rb:23]: Same comment for the other `it` blocks
https://rspec.rubystyle.guide/#it-and-specify

**Review by mwolman — CHANGES_REQUESTED** [review]

File: config/locales/es.yml [hunk 14]
```diff
@@ -17,6 +17,7 @@ es:
         other: Usuarios
   errors:
     server: 'Ha ocurrido un error'
+    network_throttling: 'Ha alcanzado el limite de solicitudes'
```
Comment (mwolman) [inline config/locales/es.yml:20]:
```suggestion
    network_throttling: 'Ha alcanzado el límite de solicitudes'
```

**Review by mwolman — APPROVED** [review]

**Review by diebas — APPROVED** [review]

**Review by lucaleivaloop — APPROVED** [review]

File: config/initializers/rack_attack.rb [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  # Throttle POST requests to /login by email param
+  #
+  # Key: "rack::attack:#{Time.now.to_i/:period}:logins/email:#{normalized_email}"
+  #
+  # Note: This creates a problem where a malicious user could intentionally
+  # throttle logins for another user and force their login requests to be
+  # denied, but that's not very common and shouldn't happen to you. (Knock
+  # on wood!)
+  throttle('logins/email', limit: 5, period: 20.seconds) do |req|
+    if req.path == '/api/v1/users/sign_in' && req.post? && req.params["user"]
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:58]: single quotes

File: spec/config/initializers/rack_attack_spec.rb [hunk 16]
```diff
@@ -0,0 +1,81 @@
+require "securerandom"
```
Comment (mconiglio) [inline spec/config/initializers/rack_attack_spec.rb:1]: single quotes

File: spec/config/initializers/rack_attack_spec.rb [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+      let(:params) do
+        {
+          user: {
+            email: "wrong+#{SecureRandom.hex(5)}@email.com",
+            password: 'wrong_password'
+          }
+        }
+      end
+
+      let(:headers) { {REMOTE_ADDR: "1.2.3.4"} }
```
Comment (mconiglio) [inline spec/config/initializers/rack_attack_spec.rb:55]: leave spaces between the brackets and the content. use single quotes

File: config/initializers/rack_attack.rb [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  # By default, Rack::Attack returns an HTTP 429 for throttled responses,
+  # which is just fine.
+  #
+  # If you want to return 503 so that the attacker might be fooled into
+  # believing that they've successfully broken your app (or you just want to
+  # customize the response), then uncomment these lines.
+  self.throttled_response = lambda do |env|
+    [
+      429,
+      {'Content-Type' => 'application/json'},
```
Comment (mconiglio) [inline config/initializers/rack_attack.rb:76]: spaces between brackets and content

**Review by mwolman — CHANGES_REQUESTED** [review]
@nbassagoda please run `rubocop -a` and fix all the issues 🙏🏻  There's a problem with the configuration of the linter that we are running in GitHub Actions and it's not blocking the PR. I already talked with @mconiglio and he's gonna handle this error 💪🏻

**Review by elestu — APPROVED** [review]

**Pull request merged**